### PR TITLE
feat(deepseek-v4): support Flash serving on SM120

### PR DESCRIPTION
**Pull request by lucifer1004**
## Summary

- add SM120 DeepSeek V4 sparse MLA and prefill indexer paths while retaining FlashMLA on its supported architectures
- add the FlashInfer CUTLASS MXFP4 MoE integration for precomputed top-k routing, including checkpoint layout conversion and SM120/SM121 capability registration
- reuse one persistent FlashInfer workspace buffer per compatible worker configuration instead of allocating one per layer or invocation
- fall back from TokenSpeed Triton symmetric-memory collectives on SM120, where that implementation is unsupported
- adapt the CuTe DSL autotuner configuration to the current FlashInfer main API

The default CUDA artifact remains targeted at SM100a/SM103a. SM120/SM121 compilation is opt-in through TOKENSPEED_CUDA_ARCH or FLASHINFER_CUDA_ARCH_LIST, so existing SM10x users do not pay the additional build cost.

The default MXFP8-activation/MXFP4-weight autotune path depends on flashinfer-ai/flashinfer#4308. That patch fixes profiler input construction; this change does not suppress any SM120 tactics or disable autotuning.

This supersedes the previously split SM120 serving work in #645 and #648. It does not supersede the P/D cache-handoff work from #660; that capability is being reworked independently in #997 against current main.

## Test Plan

- pre-commit run --all-files
- post-rebase focused kernel tests: 99 passed, 65 skipped, 3 expensive tactic-comparison tests deselected
- persistent-workspace kernel selection regression: 84 passed, 58 skipped
- post-rebase focused runtime/config/communication tests: 158 passed, 1 skipped, 13 subtests passed
- exact-shape FlashInfer fused-MoE profiling on SM120: E=128, EP=2, M=1, K=4096, N=2048, top-k=6; all GEMM1/GEMM2 profiles completed
- real two-GPU SM120 DeepSeek V4 Flash TEP2 launch through InferLab with default autotune, CUDA graph, prefill graph, OpenAI smoke request, and verified cleanup

### CODE REVIEW

File: python/tokenspeed/runtime/models/deepseek_v4.py [hunk 1]
```diff
@@ -777,6 +774,42 @@ def _deepseek_v4_indexer_topk_from_logits_prefill_op(
...
+    output: torch.Tensor,
+    topk_tokens: int,
+) -> None:
+    if _platform.is_nvidia and _platform.arch_version.major == 12:
+        if not has_indexer_topk_prefill():
+            raise RuntimeError(
+                "DeepSeek V4 prefill indexer on SM120 requires the "
+                "TokenSpeed CUDA prefill top-k op"
+            )
+        indexer_topk_prefill(
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply python/tokenspeed/runtime/models/deepseek_v4.py:790]: Fixed in 3689764b. The default CUDA artifact now includes `sm_120f` plus the exact `sm_121a` image, while `attn_res` remains explicitly limited to `sm_100a`/`sm_103a`. This covers ordinary builds without relying on the workspace override and follows the current FlashInfer SM12x targeting strategy.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/__init__.py [hunk 2]
```diff
@@ -125,6 +140,65 @@ def _topk_lens_or_count(
...
+            "FlashInfer was built without the SM120 DeepSeek V4 sparse MLA API"
+        )
+    return trtllm_batch_decode_sparse_mla_dsv4(
+        query=q,
+        swa_kv_cache=swa_kv_cache,
+        workspace_buffer=_get_dsa_sparse_workspace(q.device),
+        sparse_indices=swa_indices,
+        compressed_kv_cache=compressed_kv_cache,
+        swa_topk_lens=swa_topk_lens,
+        extra_sparse_indices=compressed_indices,
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/__init__.py:166]: No change needed here. FlashInfer current main normalizes both `[batch, q_len, topk]` and flattened `[batch * q_len, topk]` segment indices in `_normalize_sparse_mla_indices_and_lens`. This call reaches it with query shape `[tokens, 1, heads, 512]`, so the existing compressed shape `[tokens, 1, topk]` is the matching non-flattened contract and is reshaped internally. The SWA table uses the equally supported flattened form.

File: tokenspeed-kernel/test/ops/test_nvfp4_gemm_swiglu_nvfp4_quant.py [hunk 3]
```diff
@@ -299,6 +299,20 @@ def scales(rows: int, cols: int):
     ]
 
 
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="NVIDIA CUDA GPU required")
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/test/ops/test_nvfp4_gemm_swiglu_nvfp4_quant.py]: Fixed in 3689764b. The test now requires both CUDA availability and a CUDA-backed PyTorch build (`torch.version.cuda`), so ROCm no longer enters the NVIDIA-only CuTeDSL import path.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py [hunk 4]
```diff
@@ -2654,6 +2664,106 @@ def mla_decode_with_kvcache(
...
+    compressed_kv_cache: torch.Tensor | None = None,
+    compressed_indices: torch.Tensor | None = None,
+    compressed_topk_lens: torch.Tensor | None = None,
+    softmax_scale: float,
+    sinks: torch.Tensor | None = None,
+    out: torch.Tensor | None = None,
+    override: str | None = None,
+    solution: str | None = None,
+) -> torch.Tensor:
+    """Run DeepSeek V4 sparse MLA over its SWA and compressed KV pools."""
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/__init__.py]: Fixed in 3689764b. The public API docstring now documents query, cache, index and length layouts, the paired compressed-cache arguments, optional output behavior, and the returned shape.

File: python/tokenspeed/runtime/models/deepseek_v4.py [hunk 5]
```diff
@@ -2534,6 +2567,7 @@ def __init__(
                 activation="swiglu",
                 swiglu_limit=getattr(config, "swiglu_limit", None),
                 with_bias=True,
+                routing_mode="precomputed_topk",
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply python/tokenspeed/runtime/models/deepseek_v4.py]: Fixed in 22e0c36b. The model no longer constrains routing_mode. This restores the existing registry-owned selection: SM100/103 can select the FlashInfer TRT-LLM MXFP4 kernel_routing implementation, while SM120 can select the new CUTLASS precomputed_topk implementation; the selected plan remains authoritative for the TopK output format. Added a model-level regression test.

File: tokenspeed-kernel/python/setup.py [hunk 6]
```diff
@@ -70,7 +70,8 @@
 BASE_VERSION = "0.1.3"
 BACKEND_ENV = "TOKENSPEED_KERNEL_BACKEND"
 VALID_BACKENDS = {"cuda", "rocm"}
-DEFAULT_CUDA_ARCHS = ("100a", "103a")
+DEFAULT_CUDA_ARCHS = ("100a", "103a", "120f", "121a")
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/setup.py]: Fixed in 22e0c36b. _normalize_cuda_arch now preserves either explicit a or f family suffixes, including dotted forms. Added regression coverage for both 120f and 12.0f while retaining the prior behavior for unsuffixed inputs.

File: python/tokenspeed/runtime/models/deepseek_v4.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lucifer1004) [reply python/tokenspeed/runtime/models/deepseek_v4.py:790]: Reconsidered in 91960989 based on TokenSpeed current deployment baseline. The default artifact remains SM100a/SM103a; SM120/SM121 compilation is opt-in through TOKENSPEED_CUDA_ARCH or FLASHINFER_CUDA_ARCH_LIST. The SM12 support and explicit attn_res filtering remain available without expanding build cost for the primary SM10x users.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        ones = torch.ones(
+            num_local_experts,
+            dtype=torch.float32,
+            device=w.w13_weight.device,
+        )
+        w.w13_weight_global_scale = torch.nn.Parameter(ones, requires_grad=False)
+        w.w2_weight_global_scale = torch.nn.Parameter(ones.clone(), requires_grad=False)
+        swiglu_arg = getattr(w, "swiglu_arg", None)
+        w.gemm1_alpha = _expert_float_parameter(
+            w, None if swiglu_arg is None else swiglu_arg.alpha
```
[the remark this answers is not in the document]
Comment (lucifer1004) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/moe/flashinfer/cutlass_mxfp4.py]: Fixed in ea47cfea. The CUTLASS MXFP4 backend now normalizes coupled SwiGLU parameters: when beta or clamp is present and alpha is omitted, alpha defaults to the standard 1.0, which selects SwigluBias and preserves those parameters. Parameter-free standard SwiGLU remains on the existing Swiglu path. Added a focused regression test.
